src/chunkserver/chunkserver_entry.cc

```diff
@@ -0,0 +1,1578 @@
+	return &singleton;
+}
+
+std::unique_ptr<PacketStruct>
+ChunkserverEntry::createDetachedPacketWithOutputBuffer(
+    const std::vector<uint8_t> &packetPrefix) {
+	TRACETHIS();
+
+	PacketHeader header;
+	deserializePacketHeader(packetPrefix, header);
+
+	uint32_t sizeOfWholePacket = PacketHeader::kSize + header.length;
+	std::unique_ptr<PacketStruct> outPacket = std::make_unique<PacketStruct>();
+	passert(outPacket);
+
+	outPacket->outputBuffer = getReadOutputBufferPool().get(sizeOfWholePacket);
+
+	if (outPacket->outputBuffer->copyIntoBuffer(packetPrefix) !=
+	    static_cast<ssize_t>(packetPrefix.size())) {
+		if (outPacket->outputBuffer) {
+			getReadOutputBufferPool().put(std::move(outPacket->outputBuffer));
+		}
+
+		return nullptr;
+	}
+
+	return outPacket;
+}
+
+ChunkserverEntry::~ChunkserverEntry() {
+	if (sock >= 0) { tcpclose(sock); }
+	if (fwdSocket >= 0) { tcpclose(fwdSocket); }
+}
+
+void ChunkserverEntry::attachPacket(std::unique_ptr<PacketStruct> &&packet) {
+	outputPackets.push_back(std::move(packet));
+}
+
+void ChunkserverEntry::preserveInputPacket() {
+	TRACETHIS();
+	writePacket->packet = std::move(inputPacket.packet);
+}
+
+void ChunkserverEntry::createAttachedPacket(std::vector<uint8_t> &packet) {
+	TRACETHIS();
+	std::unique_ptr<PacketStruct> outpacket = std::make_unique<PacketStruct>();
+	passert(outpacket);
+
+	outpacket->packet = std::move(packet);
+	passert(outpacket->packet.data());
+
+	outpacket->bytesLeft = outpacket->packet.size();
+	outpacket->startPtr = outpacket->packet.data();
+
+	attachPacket(std::move(outpacket));
+}
+
+uint8_t *ChunkserverEntry::createAttachedPacket(uint32_t type,
+                                                uint32_t opSize) {
+	TRACETHIS();
+
+	std::unique_ptr<PacketStruct> outPacket = std::make_unique<PacketStruct>();
+	passert(outPacket);
+
+	uint32_t packetSize = opSize + PacketHeader::kSize;
+	outPacket->packet.resize(packetSize);
+	passert(outPacket->packet.data());
+
+	outPacket->bytesLeft = packetSize;
+	uint8_t *ptr = outPacket->packet.data();
+	put32bit(&ptr, type);
+	put32bit(&ptr, opSize);
+	outPacket->startPtr = outPacket->packet.data();
+
+	attachPacket(std::move(outPacket));
+
+	return ptr;
+}
+
+void ChunkserverEntry::fwdError() {
+	TRACETHIS();
+	sassert(messageSerializer != nullptr);
+	std::vector<uint8_t> buffer;
+	uint8_t status = (state == State::Connecting ? SAUNAFS_ERROR_CANTCONNECT
+	                                             : SAUNAFS_ERROR_DISCONNECTED);
+	messageSerializer->serializeCstoclWriteStatus(buffer, chunkId, 0, status);
+	createAttachedPacket(buffer);
+	state = State::WriteFinish;
+}
+
+// initialize connection to another CS
+int ChunkserverEntry::initConnection() {
+	TRACETHIS();
+	int status;
+	// TODO(msulikowski) If we want to use a ConnectionPool, this is the right
+	// place to get a connection from it
+	fwdSocket = tcpsocket();
+	if (fwdSocket < 0) {
+		safs_pretty_errlog(LOG_WARNING, "create socket, error");
+		return -1;
```

Comment:
Consider setting also an specific name for each one of these -1 returns according to the error.